test/e2e_node/dual_stack_host_ips.go

```diff
@@ -0,0 +1,286 @@
+			framework.ExpectNoError(err)
+			for _, node := range nodeList.Items {
+				if node.Name == p.Spec.NodeName {
+					nodeIPs := []string{}
+					for _, address := range node.Status.Addresses {
+						if address.Type == v1.NodeInternalIP {
+							nodeIPs = append(nodeIPs, address.Address)
+						}
+					}
+					gomega.Expect(p.Status.HostIPs).Should(gomega.Equal(nodeIPs))
+					break
+				}
+			}
+
+			ginkgo.By("deleting the pod")
+			err = podClient.Delete(ctx, pod.Name, *metav1.NewDeleteOptions(30))
+			framework.ExpectNoError(err, "failed to delete pod")
+		})
+
+		ginkgo.It("should create pod with hostNetwork, add ipv6 and ipv4 ip to host ips", func(ctx context.Context) {
+
+			podName := "pod-dualstack-host-ips"
+
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   podName,
+					Labels: map[string]string{"test": "dualstack-host-ips"},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  "dualstack-host-ips",
+							Image: imageutils.GetE2EImage(imageutils.Agnhost),
+						},
+					},
+					HostNetwork: true,
+				},
+			}
+
+			ginkgo.By("submitting the pod to kubernetes")
+			podClient := e2epod.NewPodClient(f)
+			p := podClient.CreateSync(ctx, pod)
+
+			gomega.Expect(p.Status.HostIP).ShouldNot(gomega.BeEquivalentTo(""))
+			gomega.Expect(p.Status.HostIPs).ShouldNot(gomega.BeNil())
+
+			// validate first ip in HostIPs is same as HostIP
+			framework.ExpectEqual(p.Status.HostIP, p.Status.HostIPs[0].IP)
+			if len(p.Status.HostIPs) > 1 {
+				// assert 2 host ips belong to different families
+				if netutils.IsIPv4String(p.Status.HostIPs[0].IP) == netutils.IsIPv4String(p.Status.HostIPs[1].IP) {
+					framework.Failf("both internalIPs %s and %s belong to the same families", p.Status.HostIPs[0].IP, p.Status.HostIPs[1].IP)
+				}
+			}
+
+			nodeList, err := e2enode.GetReadySchedulableNodes(ctx, f.ClientSet)
+			framework.ExpectNoError(err)
+			for _, node := range nodeList.Items {
+				if node.Name == p.Spec.NodeName {
+					nodeIPs := []string{}
+					for _, address := range node.Status.Addresses {
+						if address.Type == v1.NodeInternalIP {
+							nodeIPs = append(nodeIPs, address.Address)
+						}
+					}
+					gomega.Expect(p.Status.HostIPs).Should(gomega.Equal(nodeIPs))
+					break
+				}
+			}
+
+			ginkgo.By("deleting the pod")
+			err = podClient.Delete(ctx, pod.Name, *metav1.NewDeleteOptions(30))
+			framework.ExpectNoError(err, "failed to delete pod")
+		})
+
+		ginkgo.It("should provide hostIPs as an env var", func(ctx context.Context) {
+			podName := "downward-api-" + string(uuid.NewUUID())
+			env := []v1.EnvVar{
+				{
+					Name: "HOST_IPS",
+					ValueFrom: &v1.EnvVarSource{
+						FieldRef: &v1.ObjectFieldSelector{
+							APIVersion: "v1",
+							FieldPath:  "status.hostIPs",
+						},
+					},
+				},
+			}
+
+			expectations := []string{
+				fmt.Sprintf("HOST_IPS=%v|%v", e2enetwork.RegexIPv4, e2enetwork.RegexIPv6),
```

Comment:
shouldn't we be able to assert the order, at this point?

Reply:
The order is already ensure in the unit test
https://github.com/kubernetes/kubernetes/pull/109616/commits/3ef7c8778fedbc3069fc5ecc341bd5c5bb90239d#diff-1d1207db1b0a769c3db35da7703a0e7302f325a5efde8057a8ac7c13f1caf46dR732

This case was copied from `host_ip` and verified that it actually works
https://github.com/kubernetes/kubernetes/blob/d05b79c836c9db35ef0a60f0175b3a6bbeedfa7e/test/e2e/common/node/downwardapi.go#L86-L110